File: sdks/rust/examples/stream.rs
//! Event Stream Example
//!
//! This example demonstrates real-time event streaming.
//!
//! Usage:
//!   cargo run --example stream

use serde_json::json;
use synap_sdk::{SynapClient, SynapConfig};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize tracing
    tracing_subscriber::fmt::init();

    // Create Synap client
    let config = SynapConfig::new("http://localhost:15500");
    let client = SynapClient::new(config)?;

    tracing::info!("📡 Synap Rust SDK - Event Stream Example\n");

    // 1. Create a stream room
    tracing::info!("1. Creating stream room 'chat-room-1'");
    client
        .stream()
        .create_room("chat-room-1", Some(10000))
        .await?;
    tracing::info!("   ✅ Room created\n");

    // 2. Publish events
    tracing::info!("2. Publishing events");
    let offset1 = client
        .stream()
        .publish(
            "chat-room-1",
            "message",
            json!({"user": "alice", "text": "Hello everyone!"}),
        )
        .await?;
    tracing::info!("   Published event at offset {}", offset1);

    let offset2 = client
        .stream()
        .publish(
            "chat-room-1",
            "message",
            json!({"user": "bob", "text": "Hi Alice!"}),
        )
        .await?;
    tracing::info!("   Published event at offset {}", offset2);

    let offset3 = client
        .stream()
        .publish("chat-room-1", "join", json!({"user": "charlie"}))
        .await?;
    tracing::info!("   Published event at offset {}\n", offset3);

    // 3. Consume events from beginning
    tracing::info!("3. Consuming all events from offset 0");
    let events = client
        .stream()
        .consume("chat-room-1", Some(0), Some(10))
        .await?;
    tracing::info!("   Received {} events:", events.len());
    for event in &events {
        tracing::info!(
            "   - Offset {}: {} = {:?}",
            event.offset,
            event.event,
            event.data
        );
    }

    // 4. Consume events from specific offset
    tracing::info!("4. Consuming events from offset 1");
    let events = client
        .stream()
        .consume("chat-room-1", Some(1), Some(10))
        .await?;
    tracing::info!("   Received {} events:", events.len());
    for event in &events {
        tracing::info!(
            "   - Offset {}: {} = {:?}",
            event.offset,
            event.event,
            event.data
        );
    }

    // 5. Get room statistics
    tracing::info!("5. Getting room statistics");
    let stats = client.stream().stats("chat-room-1").await?;
    tracing::info!("   Room: {}", stats.room);
    tracing::info!("   Max offset: {}", stats.max_offset);
    tracing::info!("   Total events: {}\n", stats.total_events);

    // 6. List all rooms
    tracing::info!("6. Listing all stream rooms");
    let rooms = client.stream().list().await?;
    tracing::info!("   Rooms: {:?}\n", rooms);

    // 7. Cleanup
    tracing::info!("7. Deleting room 'chat-room-1'");
    client.stream().delete_room("chat-room-1").await?;
    tracing::info!("   ✅ Room deleted");

    tracing::info!("\n✅ Example completed successfully!");

    Ok(())
}

File: sdks/typescript/examples/stream.ts
/**
 * Stream Operations Examples
 * 
 * Demonstrates event stream operations: CREATE ROOM, PUBLISH, CONSUME, STATS, LIST
 */

import { Synap } from '../src/index';

const synap = new Synap({
  url: 'http://localhost:15500',
  timeout: 30000,
});

async function runStreamExamples() {
  console.log('🌊 === STREAM OPERATIONS EXAMPLES ===\n');

  try {
    // CREATE ROOM (rooms are auto-created on publish, but we can create explicitly)
    const roomName = `chat-room-${Date.now()}`;
    try {
      await synap.stream.createRoom(roomName);
      console.log('✅ CREATE STREAM ROOM');
    } catch (error: any) {
      // Room might already exist, that's OK
      if (!error.message?.includes('already exists')) {
        throw error;
      }
      console.log('✅ STREAM ROOM (already exists, using existing)');
    }

    // PUBLISH
    const offset1 = await synap.stream.publish(roomName, 'message.sent', {
      user: 'alice',
      text: 'Hello, world!',
      timestamp: Date.now(),
    });
    const offset2 = await synap.stream.publish(roomName, 'message.sent', {
      user: 'bob',
      text: 'Hi there!',
      timestamp: Date.now(),
    });
    console.log('✅ PUBLISH events at offsets:', offset1, offset2);

    // CONSUME
    const events = await synap.stream.consume(roomName, 'subscriber-1', 0);
    console.log('✅ CONSUME events:', events.length);

    // STATS
    const streamStats = await synap.stream.stats(roomName);
    console.log('✅ Stream Stats:', streamStats);

    // LIST ROOMS
    const rooms = await synap.stream.listRooms();
    console.log('✅ LIST ROOMS:', rooms);

    console.log('\n✅ Stream operations examples completed!');
  } catch (error) {
    console.error('❌ Error:', error);
    throw error;
  } finally {
    synap.close();
  }
}

runStreamExamples().catch(console.error);

